### PR TITLE
build a generic future<T>

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29579 build a generic future<T>**

Per #28923, this diff is to move Future<Message> to torch::utils and extend it to be Future<T>, most of implementations are copied from FutureMessage and ivalue::Future. merge ivalue::Future with Future<T> will be done separately.

The main difference between Future<T>  and FutureMessage is the error handling, instead of checking message type inside Future to handle error, this future<T> owns has_error_ and error_ states.

also this future passes value_, has_error_ and error_ states to callbacks for easily read future states.

In next diff, a torch script rpc async API will be created, before the API returns, it will create an ivalue::Future and passes it to Future<T>'s call back where state of ivalue::Future will be set.  In this way, the torch script rpc async API  can still return a ivalue::Future and call wait() to get its state appropriately afterwards.

Differential Revision: [D18263023](https://our.internmc.facebook.com/intern/diff/D18263023/)

### CODE REVIEW

**Review by pietern — COMMENTED** [review]
@zhaojuanmao How do you want to sequence the reviews?

One thought about the API -- we can make `FutureError` convertible to boolean so you can pass it as `err` argument and simply say `if (err) { // do something with err }`, instead of having an explicit bool argument to indicate if an error happened or not. What do you think?

Will this be used from the JIT as well?

**Comment by zhaojuanmao** [comment]
@pietern thanks for reviewing it! 

Sounds good to kill has_error bool in Future<T>. I will make changes accordingly. 

Yes, Zach suggested to have only one Future<T> in PyTorch code base, and JIT c10::ivalue::future should be replaced by Future<IValue> in the future. 

That would be great to review this PR first, it will be cleaner to build other PRs on top of it. Also Pritam plans to make autograd engine to be non-blocking, it may need this PR.

File: torch/csrc/utils/future.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
+#pragma once
```
Comment (mrshenli) [inline torch/csrc/utils/future.h]: Any reason for making this header only, instead of .h + .cpp?

File: torch/csrc/utils/future.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+struct TORCH_API FutureError final : public std::exception {
+  FutureError(std::string&& error_msg_) : error_msg(std::move(error_msg_)) {}
+
+  FutureError() = default;
+
+  const char* what() const noexcept override {
+    return error_msg.c_str();
+  }
+
+  std::string errMsg() const {
```
Comment (mrshenli) [inline torch/csrc/utils/future.h]: return `const std::string&` to avoid making a copy?

File: torch/csrc/utils/future.h [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  std::string error_msg;
+};
+
+// This class holds a value of type T that will be ready in the future.
+// Most implementation is copied from FutureMessage and
+// c10::ivalue::Future
+template <typename T>
+class TORCH_API Future final {
+ public:
+  using Callback = std::function<void(const T&, bool, const FutureError&)>;
```
Comment (mrshenli) [inline torch/csrc/utils/future.h]: Shall we split code copy and FutureError into separate PRs? Or is there any reason that we have do them in one?

File: torch/csrc/utils/future.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    fireCallbacks();
+    finished_cv_.notify_all();
+  }
+
+  bool completed() const {
+    return completed_;
+  }
+
+  T value() {
```
Comment (mrshenli) [inline torch/csrc/utils/future.h]: This is making a copy of the value, when do we plan to use this?

File: torch/distributed/rpc/api.py [hunk 5]
```diff
@@ -250,7 +250,7 @@ def rpc_async(to, func, args=None, kwargs=None):
     Make a non-blocking RPC call to run function ``func`` on worker ``to``. RPC
     messages are sent and received in parallel to execution of Python code. This
     method is thread-safe. This method will immediately return a
-    ``torch.distributed.FutureMessage`` that can be awaited on.
+    Future<Message> that can be awaited on.
```
Comment (mrshenli) [inline torch/distributed/rpc/api.py]: It seems `Future<Message>` is still exposed as `torch.distributed.rpc.FutureMessage` instead of `Future<Message>`, and Python developers probably don't need to know about `Future<Message>`. So probably we still say `torch.distributed.rpc.FutureMessage` here?

File: torch/csrc/distributed/rpc/init.cpp [hunk 6]
```diff
@@ -115,10 +115,12 @@ Otherwise, throws an exception.
   // is cleaned up in join_rpc(), after join_rpc(), python objects returned
   // from rpc python call can not be resolved.
   auto futureMessage =
-      shared_ptr_class_<FutureMessage>(module, "FutureMessage")
+      shared_ptr_class_<torch::utils::Future<Message>>(module, "FutureMessage")
           .def(
               "wait",
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp]: add doc for it? (can come in follow up PR)

File: torch/csrc/utils/future.h [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  void markCompleted(FutureError&& error) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    AT_ASSERT(!completed());
+    completed_ = true;
+    has_error_ = true;
+    error_ = std::move(error);
+
+    fireCallbacks();
```
Comment (jjlilley) [inline torch/csrc/utils/future.h]: lock should be dropped when firing callbacks (this is an existing problem in FutureMessage).
maybe move body of fireCallbacks() here/above to facilitate?

File: torch/csrc/utils/future.h (hunk 3 above)
Comment (jjlilley) [reply torch/csrc/utils/future.h]: wondering: maybe a template like:
    void(const T&., const FutureError* err)

where err != nullptr -> present.
or maybe two pointers? i.e. void(const T*, const FutureError *)

File: torch/csrc/utils/future.h (hunk 1 above)
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: because it has templates for class

File: torch/csrc/utils/future.h (hunk 3 above)
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: since this is generic Future<T>,  I'm copying from both FutureMessage and c10::ivalue::Future here

File: torch/csrc/utils/future.h (hunk 4 above)
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: this is to make it compatible with c10::ivalue::future, let me remove it for now. actually it will not be used

File: torch/distributed/rpc/api.py (hunk 5 above)
Comment (zhaojuanmao) [reply torch/distributed/rpc/api.py]: I will change it to expose Future<Message> then

File: torch/csrc/utils/future.h (hunk 3 above)
Comment (jjlilley) [inline torch/csrc/utils/future.h]: Wondering aloud (genuinely): do we really want exception handling part of the future framework here, rather than part of the templated T class?

In the context of Future<Message>, adding the exception handling doesn't seem to help much, since exceptions are effectively part of Message class, as-is.

And exceptions aren't strictly needed in a future impl - do we really need more than std::future<> with addCallback support here?

File: torch/csrc/utils/future.h (hunk 1 above)
Comment (mrshenli) [reply torch/csrc/utils/future.h]: We still can move implementations to cpp ([e.g.](https://github.com/pytorch/pytorch/blob/7dd06caee1dfc0d24496c8810b36ee9d7981f3be/torch/csrc/distributed/rpc/rref.cpp#L152-L155)). It's up to you. I don't have strong opinion on this. One benefit of having a separate .h would be it's clear to see what are the available APIs.

**Comment by jjlilley** [comment]
One other thought - would it be worth typedefing torch::utils::Future<Message> to FutureMessage (e.g. put this in message.h)?

This would reduce the size of the change (but not its impact), and make merge conflicts less likely.

File: torch/csrc/utils/future.h (hunk 7 above)
Comment (mrshenli) [reply torch/csrc/utils/future.h]: This might require more changes to drop the lock on `fireCallbacks()`, otherwise, `addCallback` and `fireCallbacks` can fire concurrently, creating a race on `callbacks_`

Comment (jjlilley) [reply torch/csrc/utils/future.h]: The races can be avoided because fireCallbacks() clears the callback list (mostly to guarantee that callbacks are called exactly once).

All we need is something like:

```
void makeCompleted(...) {
    std::unique_lock<std::mutex> lock(mutex_);
    ...
    std::vector<Callbacks> cbs;
    cbs.swap(callbacks_);
    lock.unlock();
    for (auto& callback : cb) {
      callback(value_);
    }
}

```
Which will be fine when used in conjunction with the existing addCallback() call, and avoids the deadlock that can occur in the (currently existing FutureMessage) code.

File: torch/csrc/utils/future.h (hunk 3 above)
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: @jjlilley that is good point. I'm motivated to add error handling here mainly because c10::ivalue::future and folly::future have this. And we need to make it compatible with c10::ivalue::future, as eventually c10::ivalue::future will implement it as Future<IValue>. 

On the other hand, I feel like it will be convenient for general cases to have this error handling, although our FutureMessage do not benefit much from it (Message could be an exception type and we have a lot of call back error handling for rref and dist autograd).  Most of classes e.g., IValue types, they do not have error types and etc, it will be hard to handle these exceptions in those cases. 

overall I prefer to have error handing in future. How do you think? @mrshenli @pritamdamania  @pietern?

Comment (jjlilley) [reply torch/csrc/utils/future.h]: Sorry, if there is an explicit goal to unify with ivalue::Future, I do see the parallelism here and the rationale for integrating the error handling. I don't have a strong objection here, and didn't mean to bring up gratuitously, I just wanted to clarify the requirements/constraints here, given that this widens the api.

(My impression was that previously there was some consensus that ivalue::Future and FutureMessage were separate enough - e.g. intrusive_ptr aspect, and easy enough to chain when we needed them to work together, that not a lot of upside in combining. But perhaps some day?)

Related question: if we go this direction, wondering if it makes more sense to use the function<void(void)> Callback being returned by ivalue::Future, rather than adding more params, since the lambda user might well just capture the future for that info?

File: torch/distributed/rpc/api.py (hunk 5 above)
Comment (pritamdamania87) [reply torch/distributed/rpc/api.py]: Ideally, we shouldn't expose `FutureMessage` or `Future<Message>` to python land. `Message` is an internal RPC concept that shouldn't leak to users. Can we just call this `Future` in python land?

File: torch/csrc/distributed/rpc/init.cpp [hunk 8]
```diff
@@ -115,10 +115,12 @@ Otherwise, throws an exception.
   // is cleaned up in join_rpc(), after join_rpc(), python objects returned
   // from rpc python call can not be resolved.
   auto futureMessage =
-      shared_ptr_class_<FutureMessage>(module, "FutureMessage")
+      shared_ptr_class_<torch::utils::Future<Message>>(module, "FutureMessage")
```
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/init.cpp]: Can we simply call this `Future` instead of `FutureMessage`? This way python users would just see a "Future" object and we don't leak "Message" to users.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]

File: torch/csrc/utils/future.h (hunk 1 above)
Comment (pietern) [reply torch/csrc/utils/future.h]: @mrshenli Only if you know exactly which functions can be specialized/instantiated up front. If the entire class is templated and it's up to the user how to use it, you can't know how it should be instantiated in advance.

Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: I still prefer to keep the implementation in header file to avoid explicit instantiation for different types, see more detailed explanation here https://stackoverflow.com/questions/495021/why-can-templates-only-be-implemented-in-the-header-file

How do you think?

File: torch/csrc/utils/future.h (hunk 3 above)
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: right, this is suggested by @pietern as well, I'm killing the flag

File: torch/csrc/utils/future.h (hunk 7 above)
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: will add this

File: torch/csrc/utils/future.h (hunk 3 above)
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: yes, we discussed with JIT team again, and the conclusion is to better have one Future in the long term, so eventually ivalue::Future should be merged. 

Ok, I will still keep error handling.

File: torch/csrc/utils/future.h [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    }
+    callbacks_.push_back(callback);
+  }
+
+ private:
+  void fireCallbacks() {
+    TORCH_CHECK(completed(), "Firing callbacks on incomplete Future.");
+    // There is no need to protect callbacks_ with the lock.
+    // Once completed_ is set to true, no one can add new callback to the list.
+    // pass value_, has_error_, error_ for callback to easily check state.
```
[the remark this answers is not in the document]
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: will do what Jeremy suggested to fire call backs without lock

File: torch/csrc/utils/future.h [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+      TORCH_CHECK(!completed());
+      completed_ = true;
+      value_ = std::move(value);
+
+      fireCallbacks();
+    }
+    finished_cv_.notify_all();
+  }
+
+  void markCompleted() {
```
[the remark this answers is not in the document]
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: good point

File: torch/csrc/utils/future.h (hunk 7 above)
Comment (zhaojuanmao) [reply torch/csrc/utils/future.h]: @jjlilley this is a good suggestion, and I'm planning to add this in a separated follow up PR to test whether it will cause flakiness or not.

File: torch/csrc/distributed/rpc/init.cpp [hunk 11]
```diff
@@ -168,12 +167,15 @@ PyObject* rpc_init(PyObject* /* unused */) {
...
-              [&](FutureMessage& fut) { return toPyObj(fut.wait()); },
-              py::call_guard<py::gil_scoped_release>());
+  auto future = shared_ptr_class_<FutureMessage>(module, "Future")
+                    .def(
+                        "wait",
+                        [&](FutureMessage& fut) { return toPyObj(fut.wait()); },
+                        py::call_guard<py::gil_scoped_release>(),
+                        R"(
+Wait on Future, it returns python object, and will throw exception if error is
+set in Future.
```
Comment (pietern) [inline torch/csrc/distributed/rpc/init.cpp]: > Wait on future to complete and return the object it completed with. If the future completes with an error, an exception is thrown.

File: torch/csrc/distributed/rpc/rref_context.cpp [hunk 12]
```diff
@@ -22,12 +22,11 @@ void RRefContext::destroyInstance(bool ignoreRRefLeak) {
   ctx.checkRRefLeaks(ignoreRRefLeak);
 }
 
-void RRefContext::handleException(const Message& message) {
-  if (message.type() == MessageType::EXCEPTION) {
+void RRefContext::handleException(const utils::FutureError* futErr) {
```
Comment (pietern) [inline torch/csrc/distributed/rpc/rref_context.cpp]: Why a const pointer?

If this is done to check if the exception is set in the first place, you could check out `c10::optional`.

File: torch/csrc/utils/future.h [hunk 13]
```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <ATen/core/ivalue.h>
+
+namespace torch {
+
+namespace utils {
+
+// FutuerError inherits from std::exception, it can return const char* or
```
Comment (pietern) [inline torch/csrc/utils/future.h]: Typo.

File: torch/csrc/utils/future.h [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+// std::string error message
+class TORCH_API FutureError final : public std::exception {
+public:
+  FutureError(std::string&& errorMsg) : errorMsg_(std::move(errorMsg)) {}
```
Comment (pietern) [inline torch/csrc/utils/future.h]: You can pass by value and then move it into place. The caller can `std::move` it into the constructor if they wish.

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit c268cb0:
* **2/2** broken upstream at merge base 717274c (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=717274c001509ffbc14189edef0f7d2ee2e5b777&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c268cb0c7935861be426da5852740e424619afff).

### 2 upstream failures recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_clang5_asan_test](https://dr.pytorch.org/build-details.html?build_id=72694046)
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=72695127)



---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

File: torch/csrc/distributed/rpc/process_group_agent.cpp [hunk 15]
```diff
@@ -377,14 +377,16 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
         if (message.isRequest()) {
           auto futureResponse = cb_->operator()(message);
           if (futureResponse->completed()) {
-            send(work.from_, std::move(*futureResponse).moveMessage());
+            send(work.from_, std::move(*futureResponse).moveValue());
```
Comment (jjlilley) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: One question that I have about the various moveValue() cases (here, 9 lines down, and parallel two cases in ThriftRpcAgent) is that we probably need to deal with the setError case.

In all these cases, it is quite possible that futureResponse was fulfilled with setError(), rather than with markCompleted(). Right now, it does seem like there's even a hasError() accessor in Future<T>?

I imagine all we need is a short helper like the following:
```
  Message packForSending(shared_ptr<FutureMessage> future) {
      if (future->hasError()) {
        return Message m(..., MessageType::EXCEPTION);
      } else {
        return std::move(*futureResponse).moveValue();
      }
  }
```

Then it something like 
```
      if (futureResponse->completed()) { 
         send(work.from_, packForSending(futureResponse));
      } 

```    
etc in the 4 rpc_agent places.

File: torch/csrc/utils/future.h [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    return value_;
+  }
+
+  T&& moveValue() && {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return std::move(value_);
+  }
+
+  void markCompleted(T value) {
+    {
```
Comment (jjlilley) [inline torch/csrc/utils/future.h]: nit: extra scope isn't really necessary here or in setError(), since the code drops the lock early now

**Review by jjlilley — COMMENTED** [review]
Nice generalization!
I just stared at this for a little while - I think I see one small set of tweaks in rpc_agent, but overall, I think this is in very reasonable shape.

File: torch/csrc/distributed/rpc/process_group_agent.cpp (hunk 15 above)
Comment (zhaojuanmao) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: @jjlilley that is great catch! I think what we need for futureMessage case is to always markCompleted(Message) then because our system may need to do some special handling based message exception type (should be rare case because we rely on error_ inside future to do error handling now).

Comment (jjlilley) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Oh, make sense - so we can choose to avoid the exception path for Future<Message>, but for generic Future<T> still have this feature.

Comment (jjlilley) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: It still might be nice to have the equivalent of a hasException() accessor like folly::Future

Comment (zhaojuanmao) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: sounds good! let me add the hasError() accessor

Comment (zhaojuanmao) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: actually we can not always markCompleted after setError, because setError is kind of markCompleted as well. 

but all the error handling are in future now, we do not need to send message with Exception type to client any more actually. I checked there is no usage in client side in our system right now.

**Comment by zhaojuanmao** [comment]
@jjlilley I updated a rebased version with addressed comments. Local tests passed. Waiting for CI.

Meanwhile would you please help taking another look at it?

File: torch/csrc/distributed/rpc/process_group_agent.cpp (hunk 15 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: So, if tests pass right now as-is, I don't see this as a blocking issue for this change.

That said, the first time somebody in request_callback_impl chooses to return a setError(), this will fail. Also note that this is server-side processing, not client side.

So, if we skip here, I will write up a followup fix in this regard

Comment (zhaojuanmao) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: @jjlilley I'm just trying to understand why it will fail if someone setError() in requestCallback?  if someone setError(), the moveValue() will just move dummy Message(), right?

**Review by jjlilley — APPROVED** [review]
Overall, I think this looks good.

Due to this change's scope, it's bound to be a magnet for merge conflicts. My feeling is that, if the tests pass cleanly, and nobody finds big red flags, we should just try to land this quickly.
Otherwise, we're likely to need to rebase over-and-over around conflicts

I will code up an rpc_agent setError-related change on top of this, mentioned in the comments above. 

Otherwise, I think it's a nice improvement to the codebase to have the generic Future<T> support available here.

File: torch/csrc/utils/future.h (hunk 1 above)
Comment (mrshenli) [reply torch/csrc/utils/future.h]: Yes, that looks good to me.

**Review by pritamdamania87 — APPROVED** [review]
looks good to me!

**Comment by zhaojuanmao** [comment]
test failures are not related, landing  soon

**Comment by ilia-cher** [comment]
"merge ivalue::Future with Future will be done separately."

Could you point me to the PR that does this:?

**Comment by zhaojuanmao** [comment]
@ilia-cher not done yet

**Comment by ilia-cher** [comment]
@zhaojuanmao do you have some PR? or at least ETA?

**Comment by zhaojuanmao** [comment]
@ilia-cher no PR yet, is it blocking some work?

**Comment by ilia-cher** [comment]
having two Future objects is not great, and to move forward we would definitely want to have a singe Future abstraction

**Comment by wanchaol** [comment]
> having two Future objects is not great, and to move forward we would definitely want to have a singe Future abstraction

@ilia-cher yeah I agree with that, that's something on our follow up/TODO list and we will spend some time to merge these two futures together in the coming months.

**Comment by ilia-cher** [comment]
Thanks, although I believe this issue is more urgent than 'coming months'

**Comment by zhaojuanmao** [comment]
Does it sound reasonable to get them merged by Mid of April?
